metals/src/main/scala/scala/meta/internal/metals/BloopServers.scala

```diff
@@ -0,0 +1,334 @@
+    sh: ScheduledExecutorService,
+    workspace: AbsolutePath,
+    client: MetalsBuildClient,
+    config: MetalsServerConfig,
+    icons: Icons
+)(implicit ec: ExecutionContextExecutorService, statusBar: StatusBar) {
+
+  def newServer(): Future[BuildServerConnection] = {
+    for {
+      (protocol, cancelable) <- callBSP()
+    } yield {
+      val tracePrinter = GlobalTrace.setupTracePrinter("BSP")
+      val launcher = new Launcher.Builder[MetalsBuildServer]()
+        .traceMessages(tracePrinter)
+        .setRemoteInterface(classOf[MetalsBuildServer])
+        .setExecutorService(ec)
+        .setInput(protocol.input)
+        .setOutput(protocol.output)
+        .setLocalService(client)
+        .create()
+      val listening = launcher.startListening()
+      val remoteServer = launcher.getRemoteProxy
+      client.onConnect(remoteServer)
+      val cancelables = List(
+        Cancelable(() => protocol.cancel()),
+        cancelable,
+        Cancelable(() => listening.cancel(true))
+      )
+      BuildServerConnection(workspace, client, remoteServer, cancelables)
+    }
+  }
+
+  private def randomPort(): Int = {
+    val s = new ServerSocket(0)
+    val port = s.getLocalPort
+    s.close()
+    port
+  }
+
+  sealed abstract class Protocol extends Cancelable {
+    import Protocol._
+    def input: InputStream = this match {
+      case Unix(socket) => socket.getInputStream
+      case Tcp(socket) => socket.getInputStream
+    }
+    def output: OutputStream = this match {
+      case Unix(socket) => socket.getOutputStream
+      case Tcp(socket) => socket.getOutputStream
+    }
+    override def cancel(): Unit = this match {
+      case Unix(socket) =>
+        if (!socket.isInputShutdown) socket.shutdownInput()
+        if (!socket.isOutputShutdown) socket.shutdownOutput()
+        socket.close()
+      case Tcp(socket) =>
+        if (socket.isClosed) {
+          socket.close()
+        }
+    }
+  }
+  object Protocol {
+    case class Unix(socket: UnixDomainSocket) extends Protocol
+    case class Tcp(socket: Socket) extends Protocol
+  }
+
+  private def callBSP(): Future[(Protocol, Cancelable)] = {
+    if (config.bloopProtocol.isTcp) callTcpBsp()
+    else callUnixBsp()
+  }
+
+  private def callTcpBsp(): Future[(Protocol, Cancelable)] = {
+    val host = "localhost"
+    val port = randomPort()
+    val args = Array(
+      "bsp",
+      "--protocol",
+      "tcp",
+      "--host",
+      host,
+      "--port",
+      port.toString
+    )
+    val cancelable = callBloopMain(args)
+    var connection: Socket = null
+    for {
+      confirmation <- waitUntilSuccess(() => {
+        try {
+          connection = new Socket(host, port)
+          true
+        } catch {
+          case NonFatal(_) =>
+            false
+        }
+      })
+    } yield {
+      if (confirmation.isYes) {
+        (Protocol.Tcp(connection), cancelable)
+      } else {
+        cancelable.cancel()
+        throw NoResponse
```

Comment:
We do a `.recover` in `MetalsLanguageServer` in case of any errors establishing a BSP connection. I estimated that wrapping things in `Either` here was not worth the additional cermony.